File: api/infrastructure/llm/openai_adapter.py
# ABOUTME: OpenAI adapter implementing ILLMProvider port for OpenAI chat completions
# ABOUTME: Converts between domain entities and OpenAI SDK types, handles streaming
from typing import AsyncIterator, List
import json
from openai import AsyncOpenAI
from ...domain.ports.llm_provider import (
    ILLMProvider, LLMEvent, LLMTextDelta, LLMToolCallDelta, LLMFinished
)
from ...domain.entities.message import Message, MessageRole, Attachment
from ...domain.entities.tool import ITool
from ...domain.exceptions import LLMProviderError, RateLimitError


class OpenAILLMAdapter(ILLMProvider):
    """OpenAI implementation of ILLMProvider port"""

    def __init__(self, api_key: str):
        self._client = AsyncOpenAI(api_key=api_key)

    async def stream_completion(
        self,
        messages: List[Message],
        tools: List[ITool],
        model: str = "gpt-4o"
    ) -> AsyncIterator[LLMEvent]:
        """Convert domain messages to OpenAI format and stream"""

        try:
            # Convert domain messages to OpenAI format
            openai_messages = self._convert_messages(messages)

            # Convert domain tools to OpenAI format
            openai_tools = [self._convert_tool(tool) for tool in tools]

            # Stream from OpenAI
            stream = await self._client.chat.completions.create(
                model=model,
                messages=openai_messages,
                tools=openai_tools if openai_tools else None,
                stream=True,
                stream_options={"include_usage": True}
            )

            # Convert OpenAI chunks to LLMEvent objects
            async for chunk in stream:
                for choice in chunk.choices:
                    # Text delta
                    if choice.delta.content:
                        yield LLMTextDelta(content=choice.delta.content)

                    # Tool call delta
                    if choice.delta.tool_calls:
                        for tool_call in choice.delta.tool_calls:
                            yield LLMToolCallDelta(
                                index=tool_call.index,
                                id=tool_call.id,
                                name=tool_call.function.name if tool_call.function else None,
                                arguments_chunk=tool_call.function.arguments if tool_call.function else None
                            )

                    # Skip finish without usage
                    if choice.finish_reason and not chunk.usage:
                        continue

                # Usage info (final chunk)
                if chunk.usage:
                    finish_reason = chunk.choices[0].finish_reason if chunk.choices else "stop"
                    yield LLMFinished(
                        finish_reason=finish_reason,
                        prompt_tokens=chunk.usage.prompt_tokens,
                        completion_tokens=chunk.usage.completion_tokens
                    )

        except Exception as e:
            # Map OpenAI errors to domain exceptions
            if "rate_limit" in str(e).lower():
                raise RateLimitError(str(e))
            raise LLMProviderError(f"OpenAI error: {str(e)}")

    def _convert_messages(self, messages: List[Message]) -> List[dict]:
        """Convert domain Message to OpenAI ChatCompletionMessageParam"""
        openai_messages = []

        for msg in messages:
            if msg.role == MessageRole.TOOL:
                # Tool response message
                openai_messages.append({
                    "role": "tool",
                    "tool_call_id": msg.tool_call_id,
                    "content": msg.content
                })
            elif msg.role == MessageRole.ASSISTANT and msg.tool_calls:
                # Assistant message with tool calls
                openai_msg = {
                    "role": "assistant",
                    "tool_calls": [
                        {
                            "id": tc.id,
                            "type": "function",
                            "function": {
                                "name": tc.name,
                                "arguments": json.dumps(tc.arguments) if not isinstance(tc.arguments, str) else tc.arguments
                            }
                        }
                        for tc in msg.tool_calls
                    ]
                }
                # Add content if present
                if msg.content:
                    openai_msg["content"] = msg.content
                openai_messages.append(openai_msg)
            else:
                # Regular message
                content_parts = [{"type": "text", "text": msg.content}]

                # Add attachments
                for attachment in msg.attachments:
                    if attachment.content_type.startswith("image"):
                        content_parts.append({
                            "type": "image_url",
                            "image_url": {"url": attachment.url}
                        })

                openai_messages.append({
                    "role": msg.role.value,
                    "content": content_parts
                })

        return openai_messages

    def _convert_tool(self, tool: ITool) -> dict:
        """Convert domain ITool to OpenAI tool format"""
        return {
            "type": "function",
            "function": {
                "name": tool.name,
                "description": tool.description,
                "parameters": tool.input_schema.model_json_schema()
            }
        }

File: src/infrastructure/adapters/ai/OpenAIAdapter.ts
// ABOUTME: Implements IAIProvider for OpenAI's GPT models
// ABOUTME: Converts domain messages to OpenAI format and streams responses

import OpenAI from 'openai';
import { IAIProvider, AICompletionRequest, AIStreamChunk, ToolDefinition } from '../../../application/ports/outbound/IAIProvider';
import { OpenAIMessageConverter } from './OpenAIMessageConverter';

export class OpenAIAdapter implements IAIProvider {
  private readonly client: OpenAI;
  private readonly messageConverter: OpenAIMessageConverter;
  private readonly defaultModel: string = 'gpt-4o';

  constructor(apiKey: string) {
    if (!apiKey) {
      throw new Error('OpenAI API key is required');
    }

    this.client = new OpenAI({ apiKey });
    this.messageConverter = new OpenAIMessageConverter();
  }

  async *streamCompletion(request: AICompletionRequest): AsyncIterable<AIStreamChunk> {
    try {
      // Convert messages to OpenAI format
      const openaiMessages = this.messageConverter.toOpenAIFormat(request.messages);
      const tools = request.tools.length > 0
        ? this.messageConverter.toOpenAITools(request.tools)
        : undefined;

      // Create streaming request
      const stream = await this.client.chat.completions.create({
        messages: openaiMessages,
        model: request.model || this.defaultModel,
        stream: true,
        tools: tools,
        temperature: request.temperature,
        max_tokens: request.maxTokens,
      });

      // Track tool calls being accumulated
      const draftToolCalls: Map<number, any> = new Map();
      const yieldedToolCalls = new Set<string>(); // Track which tool calls have been yielded

      // Process stream chunks
      for await (const chunk of stream) {
        for (const choice of chunk.choices) {
          // Handle completion
          if (choice.finish_reason === 'stop' || choice.finish_reason === 'tool_calls') {
            // If we have accumulated tool calls, yield them
            for (const [_, toolCall] of draftToolCalls) {
              // Parse arguments if not already parsed
              let args = toolCall.parsedArguments;
              if (!args && toolCall.arguments) {
                try {
                  args = JSON.parse(toolCall.arguments);
                } catch (error) {
                  console.error('Failed to parse tool arguments:', error);
                  args = {};
                }
              }

              // Only yield if we haven't already yielded this tool call
              if (!yieldedToolCalls.has(toolCall.id)) {
                yield {
                  type: 'tool_call',
                  toolCall: {
                    id: toolCall.id,
                    name: toolCall.name,
                    arguments: args || {},
                  },
                };
                yieldedToolCalls.add(toolCall.id);
              }
            }

            // Always yield usage when we have a finish reason
            // OpenAI may not send usage separately for tool calls
            yield {
              type: 'usage',
              usage: chunk.usage ? {
                promptTokens: chunk.usage.prompt_tokens,
                completionTokens: chunk.usage.completion_tokens,
                totalTokens: chunk.usage.total_tokens,
              } : {
                promptTokens: 0,
                completionTokens: 0,
                totalTokens: 0,
              },
              finishReason: choice.finish_reason,
            };
          }
          // Handle tool calls
          else if (choice.delta.tool_calls) {
            for (const toolCall of choice.delta.tool_calls) {
              const index = toolCall.index;

              if (!draftToolCalls.has(index)) {
                draftToolCalls.set(index, {
                  id: '',
                  name: '',
                  arguments: '',
                });
              }

              const draft = draftToolCalls.get(index)!;

              if (toolCall.id) {
                draft.id = toolCall.id;
              }
              if (toolCall.function?.name) {
                draft.name = toolCall.function.name;
              }
              if (toolCall.function?.arguments) {
                draft.arguments += toolCall.function.arguments;
              }

              // Try to parse accumulated arguments
              if (draft.arguments) {
                try {
                  const parsedArgs = JSON.parse(draft.arguments);
                  draft.parsedArguments = parsedArgs;
                } catch {
                  // Arguments not yet complete
                }
              }
            }
          }
          // Handle text content
          else if (choice.delta.content) {
            yield {
              type: 'text',
              content: choice.delta.content,
            };
          }
        }

        // Handle standalone usage (some models send this separately)
        if (chunk.choices.length === 0 && chunk.usage) {
          yield {
            type: 'usage',
            usage: {
              promptTokens: chunk.usage.prompt_tokens,
              completionTokens: chunk.usage.completion_tokens,
              totalTokens: chunk.usage.total_tokens,
            },
            finishReason: 'stop',
          };
        }
      }

      // Final check for completed tool calls (in case they weren't yielded yet)
      for (const [_, toolCall] of draftToolCalls) {
        // Parse arguments if not already parsed
        let args = toolCall.parsedArguments;
        if (!args && toolCall.arguments) {
          try {
            args = JSON.parse(toolCall.arguments);
          } catch (error) {
            console.error('Failed to parse tool arguments:', error);
            continue; // Skip this tool call if we can't parse it
          }
        }

        // Only yield if we haven't already yielded this tool call
        if (args && !yieldedToolCalls.has(toolCall.id)) {
          yield {
            type: 'tool_call',
            toolCall: {
              id: toolCall.id,
              name: toolCall.name,
              arguments: args,
            },
          };
          yieldedToolCalls.add(toolCall.id);
        }
      }

    } catch (error) {
      // Yield error chunk
      yield {
        type: 'error',
        error: error instanceof Error ? error.message : 'Unknown error occurred',
      };
      throw error;
    }
  }

  async getAvailableModels(): Promise<string[]> {
    try {
      const models = await this.client.models.list();
      return models.data
        .filter(model => model.id.includes('gpt'))
        .map(model => model.id);
    } catch (error) {
      console.error('Failed to fetch models:', error);
      return [this.defaultModel];
    }
  }

  async validateConnection(): Promise<boolean> {
    try {
      // Try to list models as a connectivity check
      await this.client.models.list();
      return true;
    } catch (error) {
      console.error('OpenAI connection validation failed:', error);
      return false;
    }
  }

  getProviderName(): string {
    return 'OpenAI';
  }
}